// @lona/compiler/lib/utils/printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const prettier_1 = require("prettier");
exports.builders = prettier_1.doc.builders;
function group(x) {
    if (Array.isArray(x)) {
        return exports.builders.group(exports.builders.concat(x));
    }
    return exports.builders.group(x);
}
exports.group = group;
function indent(x) {
    if (Array.isArray(x)) {
        return exports.builders.indent(exports.builders.concat(x));
    }
    return exports.builders.indent(x);
}
exports.indent = indent;
function join(x, separator) {
    if (Array.isArray(separator)) {
        return exports.builders.join(exports.builders.concat(separator), x);
    }
    return exports.builders.join(separator, x);
}
exports.join = join;
function prefixAll(x, prefix) {
    return x.map(y => exports.builders.concat([prefix, y]));
}
exports.prefixAll = prefixAll;
function print(document, options) {
    return prettier_1.doc.printer.printDocToString(document, options).formatted;
}
exports.print = print;
